src/experiments/routing/simple_graph.rs

```diff
@@ -0,0 +1,131 @@
+#[derive(Debug)]
+pub struct Graph {
+    pub vertices: Vec<Vertex>,
+    pub edges: Vec<Edge>,
+}
+
+impl Graph {
+    fn is_valid(&self) -> bool {
+        self.has_valid_vertices() && self.valid_edges()
+    }
+
+    fn has_valid_vertices(&self) -> bool {
+        only_unique_values(&self.vertices)
+    }
+
+    fn valid_edges(&self) -> bool {
+        let mut res = true;
+        for edge in self.edges.iter() {
+            res = res && self.valid_edge(edge);
+        }
+        res && only_unique_values(&self.edges)
+    }
+
+    fn valid_edge(&self, edge: &Edge) -> bool {
+        self.vertices.iter().any(|v| v.id == edge.from)
+    }
+
+    fn get_edge_weight_from_to(&self, from: i32, to: i32) -> Option<f32> {
+        let candidates: Vec<&Edge> = self.edges.iter().filter(|e| e.from == from && e.to == to).collect();
+        if candidates.len() > 1 {
+            return None;
+        }
+        match candidates.first() {
+            Some(x) => return Some(x.weight),
+            None => None
+        }
+    }
+
+    fn traverse_path_sum_weights(&self, mut path: Vec<i32>) -> f32 {
+        match path.len() {
+            0 => 0.,
+            1 => 0.,
+            _ => {
+                let from = *path.get(0).unwrap();
+                let to = *path.get(1).unwrap();
+                let edge = self.get_edge_weight_from_to(from, to).expect(&format!("There is no edge from {from} to {to}"));
+                path.remove(0);
+                return edge + self.traverse_path_sum_weights(path);
+            }
+        }
+    }
+}
+
+trait Id {
+    fn get_id(&self) -> i32;
```

Comment:
I think the convention in Rust is to not have "get" on getter methods. This would be the following instead
```
fn id(&self) -> i32
```